README.md
# PHP NodeJS Bridge

[![Latest Version on Packagist](https://img.shields.io/packagist/v/nikaia/php-node-bridge.svg?style=flat-square)](https://packagist.org/packages/nikaia/php-node-bridge)
[![Tests](https://github.com/nikaia/php-node-bridge/actions/workflows/tests.yml)

Call your NodeJS scripts from Php

## Installation

You can install the package via composer:

```bash
composer require nikaia/php-node-bridge
```

## Usage

The bridge work by executing a node script that accept piped json arbitary data, and returns json response.

> check [tests/_fixtures/ok.script.js](tests/_fixtures/ok.script.js) for a working example.


```php
use Nikaia\NodeBridge\Bridge;

try {
    $response = Bridge::create()
        ->setNode('/usr/local/bin/node')        // the path to the node (You can omit if in system path)
        ->setScript('/path/to/your/script.js')  // the path to your script 
        ->pipe(['foo' => 'bar'])                // the data to pipe to the script
        ->run();
}
catch (BridgeException $e) {
    echo $e->getMessage();
}

var_dump($response->json());   // ['foo' => 'bar']
var_dump($response->output()); // the raw output of the script {"foo":"bar"}
```
```    
``` 

```


```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Releases

This package use [semver](https://semver.org/) releases.
Releases are managed using [standard-version](https://github.com/conventional-changelog/standard-version) node package,
and requires adhering to [conventionalcommit](https://conventionalcommits.org) commit styles.

1. Implement a feature or a fix etc.
2. Use commit message like `fix: Fix an issue` or `feat: Implement a feature ...`
3. Or rewrite the commit message while squashing/closing the PR!!
4. Update your local project, checkout `main` branch
5. Run : `composer release` to generate changelog, and to tag to new release.
6. Check everything is okay.
7. Push the tag using `git push --follow-tags origin main`


> This repository is using [Semantic Pull Request bot](https://github.com/zeke/semantic-pull-requests) to enforce conventional commit message and PR titles


## Credits

- [Nassif Bourguig](https://github.com/nbourguig)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.0.0 (2022-11-16)

# Changelog

All notable changes to `nikaia/php-node-bridge` will be documented in this file.


tests/BridgeTest.php
<?php

namespace Nikaia\NodeBridge\Tests;

use Nikaia\NodeBridge\BridgeException;
use Nikaia\NodeBridge\Bridge;
use PHPUnit\Framework\TestCase;

class BridgeTest extends TestCase
{
    /** @test */
    public function it_runs_script_correctly_passing_array()
    {
        $response = Bridge::create()
            ->setScript(__DIR__ . '/_fixtures/ok.script.js')
            ->pipe(['name' => 'John Doe'])
            ->run();

        $this->assertEquals([
            'name' => 'John Doe',
            'age' => 30,
            'address' => [
                'street' => 'Main',
                'number' => 100,
            ],
        ], $response->json());
    }

    /** @test */
    public function it_runs_script_correctly_passing_json()
    {
        $response = Bridge::create()
            ->setScript(__DIR__ . '/_fixtures/ok.script.js')
            ->pipeRaw('{"name":"John Doe"}')
            ->run();

        $this->assertEquals([
            'name' => 'John Doe',
            'age' => 30,
            'address' => [
                'street' => 'Main',
                'number' => 100,
            ],
        ], $response->json());
    }

    /** @test */
    public function it_fails_when_nothing_is_piped()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessage('You must pipe a string to the nodejs script');

        (new Bridge())
            ->setScript(__DIR__ . '/_fixtures/ok.script.js')
            ->run();
    }

    /** @test */
    public function it_fails_when_script_does_not_exist()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessageMatches('/.*Error: Cannot find module.*/');

        (new Bridge())
            ->setScript(__DIR__ . '/_fixtures/does-not-exist.script.js')
            ->pipeRaw('{"name":"John Doe"}')
            ->run();
    }

    /** @test */
    public function it_fails_when_script_throws_error()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessageMatches('/.*Error: Unexpected Error!*/');

        (new Bridge())
            ->setNode('node')
            ->setScript(__DIR__ . '/_fixtures/error.script.js')
            ->pipeRaw('{"name":"John Doe"}')
            ->run();
    }
}


tests/RunnerTest.php
<?php

namespace Nikaia\NodeBridge\Tests;

use Nikaia\NodeBridge\BridgeException;
use Nikaia\NodeBridge\Runner;
use PHPUnit\Framework\TestCase;

class RunnerTest extends TestCase
{
    /** @test */
    public function it_runs_script_correctly()
    {
        $response = (new Runner())
            ->setScript(__DIR__ . '/_fixtures/ok.script.js')
            ->echoPipe('{"name":"John Doe"}')
            ->run();

        $this->assertEquals([
            'name' => 'John Doe',
            'age' => 30,
            'address' => [
                'street' => 'Main',
                'number' => 100,
            ],
        ], $response->json());
    }

    /** @test */
    public function it_fails_when_nothing_is_piped()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessage('You must pipe a string to the nodejs script');

        (new Runner())
            ->setScript(__DIR__ . '/_fixtures/ok.script.js')
            ->run();
    }

    /** @test */
    public function it_fails_when_script_does_not_exist()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessageMatches('/.*Error: Cannot find module.*/');

        (new Runner())
            ->setScript(__DIR__ . '/_fixtures/does-not-exist.script.js')
            ->echoPipe('{"name":"John Doe"}')
            ->run();
    }

    /** @test */
    public function it_fails_when_script_throws_error()
    {
        $this->expectException(BridgeException::class);
        $this->expectExceptionMessageMatches('/.*Error: Unexpected Error!*/');

        (new Runner())
            ->setScript(__DIR__ . '/_fixtures/error.script.js')
            ->echoPipe('{"name":"John Doe"}')
            ->run();
    }
}


tests/_fixtures/ok.script.js
const readInputFromPipedString = () => {
    try {
        return JSON.parse(require('fs').readFileSync(0, 'utf8'));
    } catch (e) {
        console.log('Error: invalid json input');
        process.exit(1);
    }
}

const outputToStdOut = (data) => console.log(JSON.stringify(data));


const main = () => {
    const input = readInputFromPipedString();
    outputToStdOut({
        name: input.name || 'Input issue: Not provided name',
        age: 30,
        address: {
            street: 'Main',
            number: 100
        }
    });
}
main();


src/Runner.php
<?php

declare(strict_types=1);

namespace Nikaia\NodeBridge;

use Symfony\Component\Process\Exception\ProcessFailedException;
use Symfony\Component\Process\Process;

class Runner
{
    /** @var string Full path to nodejs script file */
    protected string $script;

    /** @var string piped string */
    protected string $pipedString;

    public function __construct(
        /** @var string Node.js executable, you can specify the full path if in a custom location */
        protected string $nodeExecutable = 'node'
    ) {
    }

    /**
     * Set the nodejs js script that the runner will execute.
     */
    public function setScript(string $script): self
    {
        $this->script = $script;

        return $this;
    }

    /**
     * Set the piped string that will be sent to the nodejs script.
     */
    public function pipe(string $pipedString): self
    {
        $this->pipedString = $pipedString;

        return $this;
    }

    /**
     * Echo the passed string and pipe it to the nodejs script
     * i.e : echo '...string...' | nodejs script.js.
     */
    public function echoPipe(string $string): self
    {
        $escaped = str_replace("'", "\'", $string);

        return $this->pipe("echo '" . $escaped . "'");
    }

    /**
     * Run the nodejs script.
     */
    public function run(): Response
    {
        if (empty($this->pipedString)) {
            throw new BridgeException('You must pipe a string to the nodejs script');
        }

        $this->setUtf8Context();

        $command = "{$this->pipedString} | {$this->nodeExecutable} " . $this->script;

        $process = Process::fromShellCommandline($command);
        $process->run();

        if (! $process->isSuccessful()) {
            throw new BridgeException(
                (new ProcessFailedException($process))->getMessage()
            );
        }

        return new Response($process->getOutput());
    }

    /**
     * Fix issues where when node get executed by
     * the php process it will defaults to ascii and tries to
     * read strings in ascii, and mess everythings up.
     *
     * @see https://stackoverflow.com/a/13969829/146253
     */
    protected function setUtf8Context(): void
    {
        setlocale(LC_ALL, $locale = 'en_US.UTF-8');
        putenv('LC_ALL=' . $locale);
    }
}


src/Bridge.php
<?php

declare(strict_types=1);

namespace Nikaia\NodeBridge;

use Symfony\Component\Process\Exception\ProcessFailedException;
use Symfony\Component\Process\Process;

class Bridge
{
    /** @var string Full path to nodejs script file */
    protected string $script;

    /** @var string piped string */
    protected string $pipedString;

    public function __construct(
        /** @var string Node.js executable, you can specify the full path if in a custom location */
        protected string $nodePath = 'node'
    )
    {
    }

    public static function create(): self
    {
        return new self();
    }

    /**
     * @param string $nodePath
     * @return Bridge
     */
    public function setNode(string $nodePath): Bridge
    {
        $this->nodePath = $nodePath;

        return $this;
    }

    /**
     * Set the nodejs js script that the runner will execute.
     */
    public function setScript(string $script): self
    {
        $this->script = $script;

        return $this;
    }

    /**
     * Set the piped string that will be sent to the nodejs script.
     */
    public function setPipedString(string $pipedString): self
    {
        $this->pipedString = $pipedString;

        return $this;
    }

    /**
     * Echo the passed string and pipe it to the nodejs script
     * i.e : echo '...string...' | nodejs script.js.
     */
    public function pipeRaw(string $string): self
    {
        $escaped = str_replace("'", "\'", $string);

        return $this->setPipedString("echo '" . $escaped . "'");
    }

    /**
     * Pipe the passed array to the nodejs script after json encoding it.
     */
    public function pipe(array $input): self
    {
        return $this->pipeRaw(json_encode($input));
    }

    /**
     * Run the nodejs script.
     */
    public function run(): Response
    {
        if (empty($this->pipedString)) {
            throw new BridgeException('You must pipe a string to the nodejs script');
        }

        $this->setUtf8Context();

        $command = "{$this->pipedString} | {$this->nodePath} " . $this->script;

        $process = Process::fromShellCommandline($command);
        $process->run();

        if (!$process->isSuccessful()) {
            throw new BridgeException(
                (new ProcessFailedException($process))->getMessage()
            );
        }

        return new Response($process->getOutput());
    }

    /**
     * Fix issues where when node get executed by
     * the php process it will defaults to ascii and tries to
     * read strings in ascii, and mess everythings up.
     *
     * @see https://stackoverflow.com/a/13969829/146253
     */
    protected function setUtf8Context(): void
    {
        setlocale(LC_ALL, $locale = 'en_US.UTF-8');
        putenv('LC_ALL=' . $locale);
    }
}
